handler.go

```diff
@@ -0,0 +1,614 @@
+	dns.RcodeBadKey:         "Key not recognized",
+	dns.RcodeBadTime:        "Signature out of time window",
+	dns.RcodeBadMode:        "Bad TKEY Mode",
+	dns.RcodeBadName:        "Duplicate key name",
+	dns.RcodeBadAlg:         "Algorithm not supported",
+	dns.RcodeBadTrunc:       "Bad Truncation",
+	dns.RcodeBadCookie:      "Bad/missing Server Cookie",
+}
+
+// ServeDNS implements the plugin.Handler interface.
+func (resolver *OCPDNSNameResolver) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	state := request.Request{W: w, Req: r}
+
+	// Record response to get status code and size of the reply.
+	rw := dnstest.NewRecorder(w)
+
+	// Get the response for the DNS lookup from the plugin chain.
+	status, err := plugin.NextOrFailure(resolver.Name(), resolver.Next, ctx, rw, r)
+
+	// Get the DNS name from the DNS lookup request.
+	qname := strings.ToLower(state.QName())
+
+	var regularDnsInfo, wildcardDnsInfo namespaceDNSInfo
+	var regularDNSExists, wildcardDNSExists bool
+
+	// Check if the query was for a wildcard DNS name or a regular DNS name.
+	if isWildcard(qname) {
+		// Get the wildcard DNS name info, if it exists.
+		wildcardDnsInfo, wildcardDNSExists = resolver.wildcardDNSInfo[qname]
+	} else {
+		// Get the regular DNS name info, if it exists.
+		regularDnsInfo, regularDNSExists = resolver.regularDNSInfo[qname]
+
+		// Get the corresponding wildcard DNS name for the reguar DNS name.
+		wildcard := getWildcard(qname)
+		// Get the wildcard DNS name info, if it exists.
+		wildcardDnsInfo, wildcardDNSExists = resolver.wildcardDNSInfo[wildcard]
+	}
+
+	// If neither regular DNS name info nor wildcard DNS name info exists for the DNS name
+	// then return the response received from the plugin chain.
+	if !regularDNSExists && !wildcardDNSExists {
+		return status, err
+	}
+
+	// Check if the DNS lookup is unsuccessful or an error is encountered during the lookup.
+	if status != dns.RcodeSuccess || err != nil {
+		// WaitGroup variable used to wait for the completion of update of DNSNameResolver CRs
+		// corresponding to the regular and the wildcard DNS names.
+		var wg sync.WaitGroup
+
+		// If regular DNS name info exists then update the corresponding DNSNameResolver CR for
+		// the DNS lookup failure.
+		if regularDNSExists {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				resolver.updateResolvedNamesFailure(ctx, regularDnsInfo, qname, status)
+			}()
+		}
+
+		// If wildcard DNS name info exists then update the corresponding DNSNameResolver CR for
+		// the DNS lookup failure.
+		if wildcardDNSExists {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				resolver.updateResolvedNamesFailure(ctx, wildcardDnsInfo, qname, status)
+			}()
+		}
+
+		// Wait for the goroutines to complete.
+		wg.Wait()
+
+		// Return the response received from the plugin chain.
+		return status, err
+	}
+
+	// Get the IP addresses and the corresponding TTLs in a map. Only A and AAAA type DNS records
+	// are considered.
```

Comment:
CNAME records don't return any IP addresses. They return another DNS name which the client has to again lookup. When the second lookup happens the record returned will be either A or AAAA record containing the IP addresses. The plugin will pickup the IP addresses then if the DNS name matches any DNSNameResolver CR.

Reply:
> When the second lookup happens the record returned will be either A or AAAA record containing the IP addresses.

Who does the second lookup: the user's client or the plugin? What if the CNAME points to a DNS name not mentioned in `DNSNameResolver` CRs?

Reply:
> Who does the second lookup: the user's client or the plugin?

The client does the second lookup.

> What if the CNAME points to a DNS name not mentioned in `DNSNameResolver` CRs?

Then the plugin will not pick up the IP addresses. For EgressFirewall to work correctly, the users have to create rules containing DNS names pointed by the CNAME record. Otherwise, it'll not allow the connection to the DNS name.

Reply:
> > > What if the CNAME points to a DNS name not mentioned in DNSNameResolver CRs?
> > 
> > 
> > Then the plugin will not pick up the IP addresses. For EgressFirewall to work correctly, the users have to create rules containing DNS names pointed by the CNAME record. Otherwise, it'll not allow the connection to the DNS name.
> 
> What if the CNAME record's target cannot be guessed? For instance the apps wildcard record created by OpenShift targets a CLB's DNS name:
> 
> ```
> *.apps.mycluster.devcluster.openshift.com -> ac3b1e3613d7e4c789abbaeeeeeeeeee-2063721285.us-east-2.elb.amazonaws.com.
> ```

Normally a CNAME record's target should not guessed when the EgressFirewall rule is created, rather the CNAME record should be looked up to get the target and the rule for the target should be added, if it's an A or AAAA record. So, it is the user's responsibility to add the rules corresponding to the DNS name referred by the CNAME records, even if the record is not human readable.

Reply:
Sure, this can be added to the docs. The current implementation of EgressFirewall also only uses A or AAAA records. Maybe supporting CNAME records can be a future feature that can be added if the customers request for it.